File: python/03_003.py
def find_maximum(arr):
  """
  Finds the maximum value in an array/list.
  
  Args:
     arr: A list of comparable elements
  
  Returns:
     The maximum value found in the array
  
  Raises:
     IndexError: If the input array is empty
  """
  # Initialize max with first element
  max_value = arr[0]
  
  # Iterate through each element
  for num in arr:
     # If current element is greater
     if num > max_value:
        # Update max_value
        max_value = num
  
  # Return the maximum value
  return max_value

# Example usage
numbers = [3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5]
print(find_maximum(numbers))  # Output: 9

File: data structures and algorithms/java/03_003.java
public class LinearTime {
  public static int findMaximum(int[] arr) {
     // Check for empty array
     if (arr.length == 0) {
        throw new IllegalArgumentException(
           "Array cannot be empty");
     }
     
     int maxValue = arr[0];
     for (int num : arr) {
        if (num > maxValue) {
           maxValue = num;
        }
     }
     return maxValue;
  }

  public static void main(String[] args) {
     int[] numbers = {3, 1, 4, 1, 5, 9, 2, 6, 5, 3, 5};
     System.out.println("Maximum value: " + 
        findMaximum(numbers));  // Output: 9
  }
}